lib/puppetserver/ca/action/list.rb

```diff
@@ -79,41 +88,62 @@ def run(input)
           requested, signed, revoked = separate_certs(all_certs)
           missing = certnames - all_certs.map { |cert| cert['name'] }
 
-          (all || certnames.any?) \
-            ? output_certs_by_state(requested, signed, revoked, missing)
-            : output_certs_by_state(requested)
+          (all || certnames.any?)\
+            ? output_certs_by_state(all, output_format, requested, signed, revoked, missing)
+            : output_certs_by_state(all, output_format, requested)
 
           return missing.any? \
             ? 1
             : 0
         end
 
-        def output_certs_by_state(requested, signed = [], revoked = [], missing = [])
+        def output_certs_by_state(all, output_format, requested, signed = [], revoked = [], missing = [])
+          grouped_cert = {}
           if revoked.empty? && signed.empty? && requested.empty? && missing.empty?
             @logger.inform "No certificates to list"
             return
           end
 
           unless requested.empty?
-            @logger.inform "Requested Certificates:"
-            output_certs(requested)
+            if output_format
+              all ? grouped_cert["requested"] = requested : @logger.inform(requested.to_json)
+            else
+              @logger.inform "Requested Certificates:"
+              output_certs(requested)
+              end
           end
 
           unless signed.empty?
-            @logger.inform "Signed Certificates:"
-            output_certs(signed)
+            if output_format
+              all ? grouped_cert["signed"] = signed : @logger.inform(signed.to_json)
+            else
+              @logger.inform "Signed Certificates:"
+              output_certs(signed)
+            end 
           end
 
           unless revoked.empty?
-            @logger.inform "Revoked Certificates:"
-            output_certs(revoked)
+            if output_format
+              all ? grouped_cert["revoked"] = revoked : @logger.inform(revoked.to_json)
+            else
+              @logger.inform "Revoked Certificates:"
+              output_certs(revoked)
+            end
           end
 
           unless missing.empty?
-            @logger.inform "Missing Certificates:"
-            missing.each do |name|
-              @logger.inform "    #{name}"
-            end
+            if output_format
+              all ? grouped_cert["missing"] = missing : @logger.inform(missing.to_json)
+            else
+              @logger.inform "Missing Certificates:"
+              missing.each do |name|
+                @logger.inform "    #{name}"
+              end
+            end 
+          end
+
+          if all
```

Comment:
Will this always print a potentially empty json object?

Reply:
Ahhh. I forgot to add in the check for the `output_format`. Thanks for catching that.